ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,398 @@
+-- the necessary inpure machinery to retrieve their inputs. These rules are
+-- translated as verbatim as possible from:
+--
+-- https://github.com/cardano-foundation/CIPs/blob/master/CIP-0140/README.md#rules-for-voting-in-a-round
+module Ouroboros.Consensus.Peras.Voting.Rules
+  ( PerasVotingView (..)
+  , mkPerasVotingView
+  , isPerasVotingAllowed
+  , PerasVotingRule (..)
+  , VoteReason (..)
+  , NoVoteReason (..)
+  , perasVR1A
+  , perasVR1B
+  , perasVR2A
+  , perasVR2B
+  , perasVR1
+  , perasVR2
+  , perasVotingRules
+  )
+where
+
+import Cardano.Slotting.Slot (WithOrigin)
+import Data.Bifunctor (bimap)
+import Data.Typeable (Typeable)
+import Ouroboros.Consensus.Block (WithOrigin (..))
+import Ouroboros.Consensus.Block.Abstract
+  ( GetHeader (..)
+  , Header
+  , SlotNo (..)
+  , StandardHash
+  , castPoint
+  , succWithOrigin
+  )
+import Ouroboros.Consensus.Block.SupportsPeras
+  ( HasPerasCertRound (..)
+  , PerasRoundNo (..)
+  , ValidatedPerasCert
+  , getPerasCertBoostedBlock
+  , getPerasCertRound
+  , onPerasRoundNo
+  )
+import Ouroboros.Consensus.BlockchainTime.WallClock.Types
+  ( WithArrivalTime (..)
+  )
+import Ouroboros.Consensus.HardFork.History.EraParams (fromPerasEnabled)
+import qualified Ouroboros.Consensus.HardFork.History.Qry as HF
+import qualified Ouroboros.Consensus.HardFork.History.Summary as HF
+import Ouroboros.Consensus.Peras.Params
+  ( PerasBlockMinSlots (..)
+  , PerasCertArrivalThreshold (..)
+  , PerasCooldownRounds (..)
+  , PerasIgnoranceRounds (..)
+  , PerasParams (..)
+  )
+import Ouroboros.Consensus.Util.Pred
+  ( Evidence
+  , Explainable (..)
+  , ExplanationMode (..)
+  , Pred (..)
+  , evalPred
+  )
+import Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+
+{-------------------------------------------------------------------------------
+  Voting interface
+-------------------------------------------------------------------------------}
+
+-- | Interface needed to evaluate the Peras voting rules
+data PerasVotingView cert = PerasVotingView
+  { perasParams :: PerasParams
+  -- ^ Peras protocol parameters.
+  , currRoundNo :: !PerasRoundNo
+  -- ^ The current Peras round number.
+  , latestCertSeen :: !(WithOrigin cert)
+  -- ^ The most recent certificate seen by the voter.
+  , latestCertOnChain :: !(WithOrigin cert)
+  -- ^ The most recent certificate present in some chain.
+  , certRoundStart :: cert -> SlotNo
+  -- ^ Get the slot number at the start of the Peras round of a certificate.
+  , arrivalSlot :: cert -> SlotNo
+  -- ^ Get the arrival slot number of a certificate
+  , candidateExtendsCert :: cert -> Bool
+  -- ^ Does the candidate block extend the one boosted by a certificate?
+  }
+
+-- | Construct a 'PerasVotingView'.
+--
+-- NOTE: this assumes that the client code computes all the needed inputs
+-- within the same STM transaction, or the results may be inconsistent.
+mkPerasVotingView ::
```

Comment:
At this point I still don't see what is the "candidate block" wrt. this function.